### DISCUSSION TITLE
/hyndsight/tscvexample/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#89] [Announcements]
Time series cross-validation: an R example 
 https://robjhyndman.com/hyndsight/tscvexample/

**Comment by robjhyndman** [comment]
```header
Originally posted on 2011-09-07 02:22:51
````
Udo Sglavo shows how to do this in SAS here: [Part
1](http://blogs.sas.com/content/forecasting/2011/09/02/guest-blogger-udo-sglavo-on-cross-validation-using-sas-forecast-server-part-1-of-2/ &quot;http://blogs.sas.com/content/forecasting/2011/09/02/guest-blogger-udo-sglavo-on-cross-validation-using-sas-forecast-server-part-1-of-2/&quot;)
and [Part
2](http://blogs.sas.com/content/forecasting/2011/09/06/guest-blogger-udo-sglavo-on-cross-validation-using-sas-forecast-server-part-2-of-2/ &quot;http://blogs.sas.com/content/forecasting/2011/09/06/guest-blogger-udo-sglavo-on-cross-validation-using-sas-forecast-server-part-2-of-2/&quot;).

**Comment by robjhyndman** [comment]
```header
Originally posted by Zachary Mayer on 2011-11-21 16:59:28
````
I parallelized this algorithm using foreach:\
[http://moderntoolmaking.blo\...](http://moderntoolmaking.blogspot.com/2011/11/functional-and-parallel-time-series.html &quot;http://moderntoolmaking.blogspot.com/2011/11/functional-and-parallel-time-series.html&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2011-11-21 23:13:09
````
Nice work! How fast is it compared to my direct approach?

**Reply by robjhyndman** [reply]
```header
Originally posted by Zachary Mayer on 2011-11-22 16:34:28
````
This was all run on my laptop, under windows.  I used the doRedis
backend and all 4 cores for the parallel runs.\
Linear model (non-parallel): Rob: 3.51, Zach: 2.06\
Arima model: Rob: 184 Zach: 100\
ETS model: Rob: 535 Zach: 410\
It looks like I\'m getting a slightly different answer than you for the
error at the 1st forecast horizon.  Other than that, the results appear
identical.

**Reply by robjhyndman** [reply]
```header
Originally posted by Zachary Mayer on 2011-11-22 16:40:20
````
I also ran the Arima model sequentially under my code, and it took 188
seconds.

**Reply by robjhyndman** [reply]
```header
Originally posted by Zachary Mayer on 2012-07-29 19:58:52
````
I\'ve posted this code to github, and will gladly accept
patches/comments:\
[https://github.com/zachmaye\...](https://github.com/zachmayer/cv.ts)

**Comment by robjhyndman** [comment]
```header
Originally posted by Zachary Mayer on 2011-11-28 18:55:20
````
In the first block of code, you setup the for loop as follows: \'for(i
in 1:(n-k-1))\'.  This seems to skip the last possible value of xnext,
which is a vector of length 1, with a value of 19.43174, for June 2008.
 Why note use \'for(i in 1:(n-k))\' to setup your for loop, and include
this last value in your test set?  (If you change the
MAE matrices to matrix(NA,n-k,12) the code runs fine)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2011-11-28 23:42:03
````
Thanks Zach. I\'m not sure why I left that one out. In any case, I\'ve
now updated the code as you suggest.

**Comment by robjhyndman** [comment]
```header
Originally posted by Anusha Anusha on 2013-01-27 14:32:24
````
Dear Sir, Thank you for a useful post. Would you please give a reference
for the above method of cross validating time series? Also I have not
found many papers discussing time series cross validation. Is the method
suggested above widely acceptable or most used method for carrying out
cross validation for time series and are there other alternatives that
must be considered along with the above method?

Thanking you\
Anusha

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-01-29 01:38:16
````
It is widely used in forecasting research, but is not normally called
cross-validation. I was pointing out that this approach is analogous to
cross-validation used for cross-sectional data. For references, see my
book at
[http://otexts.com/fpp/2/5/](http://otexts.com/fpp/2/5/)
or [Fildes (1992) vol 8,
p81-98](http://www.sciencedirect.com/science/article/pii/016920709290009X).

**Comment by robjhyndman** [comment]
```header
Originally posted by Matthew Dickinson on 2013-02-06 22:13:26
````
Thank you for the demonstration. How do you integrate the
cross-validation procedures when batch forecasting (as you described in
[http://robjhyndman.com/rese\...](http://robjhyndman.com/researchtips/batch-forecasting/ &quot;http://robjhyndman.com/researchtips/batch-forecasting/&quot;)
)? In other words, is there a way to use auto.arima or forecast to
select the best fitting model while incorporating the cross-validation
techniques?

Sorry if this is a duplicate comment, my first one seemed to have
vanished.

Thank you,

MD

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-02-09 22:22:10
````
auto.arima does not use cross-validation. You would have to write
something similar to auto.arima which uses cv to select the model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Roy Sanderson on 2013-02-27 17:36:19
````
Dear Rob

Thanks for the excellent description. I\'ve been wondering whether this
will help me crack an apparently simple problem: I\'ve a seasonal Arima
model (2,1,1)x(1,0,1) 24hrs, with hrly records over 78 days, and 5
significant xreg variables. I\'m trying to determine the &quot;effect size&quot;
of each xreg variable, and have been setting 4 of them to their means
whilst leaving 1 at is original, calculating predicted values, and doing
this in turn for each xreg. Is this correct (whilst it\'s the usual way
for lm, perhaps not for ts)? Even if it is correct, I\'ve no idea how to
calculate 95% prediction intervals for xreg effect sizes from an Arima.
So I was wondering whether to calculate lots of Arimas, in a
\'leave-k-out\' approach for all but a subset of the data, and
forecasting values for each small subset. As this will take a very long
time to run on my slow PC I thought I\'d better check with you first!\
Many thanks for any hints.

Roy

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-02-28 04:19:07
````
Your approach will give you estimated effect sizes. However, I don\'t
know what you mean by prediction intervals for effect sizes. A
prediction interval is for an unobserved random variable, whereas an
effect size is not random. Perhaps you mean a confidence interval? In
that case, the ci can be estimated as x times the ci for beta.

**Reply by robjhyndman** [reply]
```header
Originally posted by Roy Sanderson on 2013-03-02 17:21:13
````
Dear Rob\
Many thanks for your reply - apologies, I should have said 95% CI. What
I\'m after is plots similar to those obtained via the allEffects
function, except for the xreg variables used in an arima rather than a
lm. Many thanks if you think the method I\'ve suggested should be OK.\
Roy

**Comment by robjhyndman** [comment]
```header
Originally posted by Isabella Ghement on 2013-07-18 05:32:29
````
Dear Prof. Hyndman,

In practice, would the entire time series be used to guess that the
ARIMA model should be specified as ARIMA(p,d,q)(P,D,Q)s? where p=3, d=0,
q=1, P=0, D=1,Q=1 and s=12?

Once the structure of the ARIMA model is identified from the entire time
series, is this same structure used when producing rolling-window
forecasts?

Thank you in advance for considering my question.\
Isabella

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-07-18 06:03:43
````
You would normally be trying several models and selecting the best based
on a cross-validation procedure.

**Reply by robjhyndman** [reply]
```header
Originally posted by Isabella Ghement on 2013-07-19 05:14:17
````
Thank you so much for your prompt response. I am still unclear on how
this is supposed to work in practice, so I will elaborate below on my
confusion.

Let\'s say that I would like to predict the unknown value of a future
observation y\_{t+1} on the basis of the observations y\_{1}, \...,
y\_{n}.

Assuming that y\_{1}, \..., y\_{n} follow an ARIMA(p,d,q) process, I
would proceed to identify sensible values for p,d,q from the data
y\_{1}, \..., y\_{n} . For simplicity, let\'s denote these values by
p\*, d\*, q\*.

Given the values p\*, d\*, q\*, I would estimate the parameters of the
ARIMA(p\*,d\*,q\*) process and use them to produce a forecast of the
future observation y\_{t+1}.

Next, I would try to produce a measure of accuracy to accompany this
forecast using the &quot;rolling forecast origin&quot; procedure described at
[http://otexts.com/fpp/2/5/](http://otexts.com/fpp/2/5/).

To this end, I would fit &quot;forecast&quot; ARIMA models to the following
rolling training sets (where m \&lt; n):

1\) y\_{1}, \..., y\_{m};\
2) y\_{1}, \..., y\_{m+1};\
3) y\_{1}, \..., y\_{m+2};\
etc.

My confusion is this: What is the connection between these &quot;forecast&quot;
ARIMA models and the ARIMA(p\*,d\*,q\*) model identified from the entire
series?

Do we assume that the &quot;forecast&quot; ARIMA models have the same order
structure as the ARIMA(p\*,d\*,q\*) model?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-07-19 06:26:30
````
Cross-validation is both a method of measuring accuracy \*and\* a method
for choosing a model. The model you finally use for forecasting is the
one that gives the best cross-validation accuracy.

**Reply by robjhyndman** [reply]
```header
Originally posted by qoheleth on 2014-03-04 23:16:33
````
This might or might not be Isabella\'s question. You don\'t normally
start the modelling with CV. You would have first made a rough guess on
(p,d,q) maybe using your auto.arima function or some other procedure
that uses all the data. Then you feed this into a CV routine, possibly
with some variants of this model for comparison. The question is, in
this case, the test set is no longer unseen, so maybe the test error
will still be biased downwards? (this is a question)

**Comment by robjhyndman** [comment]
```header
Originally posted by Maria Froes on 2013-08-15 18:30:17
````
Dear Prof. Hyndman, one thing I don\'t understand is, in the following
command:

st \&lt;- tsp(a10)\[1\]+(k-2)/12

why do you subtract k-2?\
Thank you

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-15 22:49:09
````
First, you copied it incorrectly. The post said

st \&lt;- tsp(a10)\[1\]+(k-1)/12

This is the time that the shortest possible training set ends. We want k
values in the training set, so there are k-1 values after the starting
time.

**Reply by robjhyndman** [reply]
```header
Originally posted by Maria Froes on 2013-08-15 23:24:19
````
Got it! Thank you!\
But I think I haven\'t copied it incorrectly. In your first set of code
(line 6) you have k-2. But I understood the k-1. Thank you again.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-15 23:28:15
````
Sorry. I was looking at the last set. In that first set, I subtracted
one more because I added i/12 in the loop, starting with i=1.

**Reply by robjhyndman** [reply]
```header
Originally posted by Michael Lovell on 2013-11-11 22:23:30
````
Would that essentially be like using the leave-one-out method of
cross-validation such as in the following example?\...

[http://jaquesgrobler.github\...](http://jaquesgrobler.github.io/Online-Scikit-Learn-stat-tut/modules/generated/sklearn.cross_validation.LeaveOneOut.html#sklearn.cross_validation.LeaveOneOut &quot;http://jaquesgrobler.github.io/Online-Scikit-Learn-stat-tut/modules/generated/sklearn.cross_validation.LeaveOneOut.html#sklearn.cross_validation.LeaveOneOut&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-11-12 00:39:24
````
No. The LOO CV described there is for non-time-series data. See
[http://robjhyndman.com/hynd\...](http://robjhyndman.com/hyndsight/crossvalidation/)
for discussion of that problem.

**Reply by robjhyndman** [reply]
```header
Originally posted by Michael Lovell on 2013-11-12 20:22:36
````
Thank you for your explanation. However, per this illustration:
[http://i.imgur.com/Som6cwz.png](http://i.imgur.com/Som6cwz.png)

if the dataset in the illustration was time series data and was sorted
from past to present from left to right, I\'m still a bit confused on
how Leave-One-Out wouldn\'t be identical to your explanation on proper
Time Series CV.\
Thanks again!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-11-12 21:24:07
````
No. &quot;Leave one out&quot; means leave ONE observation out. Time series CV
means leave ALL future observations out.

**Comment by robjhyndman** [comment]
```header
Originally posted by Anon on 2013-10-29 21:33:49
````
How would this approach be applied in a nested fashion (i.e. what is the
time series equivalence to nested cross validation)?

**Comment by robjhyndman** [comment]
```header
Originally posted by brian on 2013-12-27 20:59:09
````
I have recently tried something similar. I\'ve set up a loop to fit a
model on a rolling subset (fixed number of obs). I then compute a 1-step
ahead forecast. Collecting the individual forecasts into a data frame, I
then regress the actual results against the 1-step ahead forecasts. How
should I interpret the parameters of this second regression? If I want
to make a 1-step ahead forecast, am I better off using the raw model
predictions (from the latest rolling regression) or plug in the model
predictions into my second regression?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-12-30 03:45:03
````
The coefficients measure bias in the forecasts. Unless something is very
wrong, the intercept should be insignificantly different from zero and
the slope should be insignificantly different from one. Assuming that is
true, just use the original predictions.

**Comment by robjhyndman** [comment]
```header
Originally posted by qoheleth on 2014-03-04 23:15:49
````
If sample size n =100, but frequency is 35, would it make sense align h
to be 35? i.e. use first 30 to be test 31\...65, then use first 65 to
test 66..100? The problem is of course the error becomes very variable.
Or would you ignore frequency when doing CV?

**Comment by robjhyndman** [comment]
```header
Originally posted by ubergooroo on 2014-05-09 15:57:29
````
Can I choose a different value for K? what if my entire data length is
just 60 periods?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-05-10 02:15:54
````
Choose whatever value of k you think is reasonable.

**Comment by robjhyndman** [comment]
```header
Originally posted by Affy on 2014-10-28 09:21:15
````
Dear Professor Hyndman,

Thank you very much for your very useful blog.

I\'m doing one step ahead forecast for time series data set using a
univariate regression model. I tried to adjust these codes for my work.
I have 232 observations for each variable.I consider 152 as in-sample
and 80 observations as out-of-sample. When I run my program
surprisingly, I receive 80 list that each list includes 152
forecasts.However, I expect to get just 80 forecast as output of my
program. I\'ve checked and really cannot understand why it happens!

#Out-of-Sample forecast\
library(forecast)\
h\&lt;-1\
train= window(data, start=1995,end=2007.66)\
test=window(data,start=2007.66)\
n=nrow(test) #number of out-of-sample observation which is 80\
fc = list()\
fit=lm(train\[,2\]\~train\[,3\])

for(i in 1:n)\
{\
x=window(data, start= 1995, end=2007.66+(i-1)/12)

refit= lm(x\[,2\]\~x\[,3\])

fc\[\[i\]\]=forecast(refit,newdata=x,h)\$mean\
}

Also, I don\'t know how can I change the code to regress x\[,2\] on lag
of x\[,3\] and use that for my forecast.I really tried to solve these
problems but I couldn\'t. I really appreciate if you help me to find the
problem with my codes.

**Comment by robjhyndman** [comment]
```header
Originally posted by Jake on 2015-05-28 17:03:51
````
Dear Dr. Hyndman,

When I tried to do the one-step forecast method using your code listed
above, I believe I found an error. The second forecast runs into a
problem trying to compute the mae. Specifically at this point

mae2\[i,1:length(xnext)\] \&lt;- abs(fcast2\[\[\'mean\'\]\]-xnext)

At this point R gives the right values, but doesn\'t spit them out in a
time series format, which throws off the the for loop and results in an
error. Any idea on how to fix this?

Thanks,\
Jake

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-05-29 01:23:42
````
Sorry. I\'ve now fixed the code.

**Comment by robjhyndman** [comment]
```header
Originally posted by Trang Nguyen on 2016-06-08 13:57:27
````
Dear Prof. Hyndman,

Could you please advise me how to adjust your code (first block) to
quarterly time series? I tried as following, but something is not
correct because I got warning that end value of xnext is extending. I
need the code to proceed my Master Thesis, so I would appreciate very
much your help.

This is my adjusted code:

data \&lt;- read.xlsx(&quot;Data/MainVariables_2010_wo time.xlsx&quot;, 1)\
GVA \&lt;- data\[,1\]\
GVA_ts \&lt;- ts(log(GVA), start=c(1991,1), end=c(2010,4), frequency=4)\
GVA_tf \&lt;- diff(GVA_ts, differences=ndiffs(GVA_ts))

k \&lt;- 60 \# minimum data length for fitting a model\
n \&lt;- length(GVA_tf)\
mae \&lt;- matrix(NA,n-k,4)\
st \&lt;- tsp(GVA_tf)\[1\]+(k-2)/4

for(i in 1:(n-k)) {\
xshort \&lt;- window(GVA_tf, end=st + i/4)\
xnext \&lt;- window(GVA_tf, start=st + (i+1)/4, end=st + (i+4)/4, extend =
TRUE)\
fit \&lt;- ar.ols(xshort, aic = TRUE, order.max = 4, demean = FALSE)\
fcast \&lt;- forecast(fit, h=4)\
mae\[i,1:length(xnext)\] \&lt;- abs(fcast\[\[\'mean\'\]\]-xnext)\
}

This is my results of mae:\
\&gt; tail(mae)\
\[,1\] \[,2\] \[,3\] \[,4\]\
\[14,\] 0.017489173 0.014937793 0.028920995 0.018692716\
\[15,\] 0.004649904 0.014389252 0.013492540 0.008615116\
\[16,\] 0.012805091 0.012952909 0.008431320 0.005965961\
\[17,\] 0.008439359 0.006840779 0.005405601 NA\
\[18,\] 0.003659734 0.004208891 NA NA\
\[19,\] 0.002812577 NA NA NA

Thanks and best regards,\
Trang Nguyen

**Reply by robjhyndman** [reply]
```header
Originally posted by christopher meier on 2016-08-02 14:27:24
````
Hi Trang,

Where you able to get this to work? I am looking to work with this code
as well for greater flexibility in modeling it provides compared caret()
and [ts.cv](http://ts.cv &quot;ts.cv&quot;)() since I can
blend techniques. I am working with hourly data and am unable to tell if
all the hard coded \'12\'s can be set as a single parameter.

**Comment by robjhyndman** [comment]
```header
Originally posted by mukund on 2016-12-31 17:17:19
````
When should one use the different types of ARIMA model as mentioned
below:

1.Estimate the model order in the training data set and use the same
order to forecast future values (updating the parameter estimates)\
2.Use a rolling window (e.g. 30 day )to make a new forecast by
estimating model order and parameters everytime to make a new forecast.\
.\
I tried the above strategies to forecast the daily downtimes of the
machine as a percentage of scheduled time in manufacturing. I used the
cross-validation methods and found that the rolling window with the new
estimation of model order (p,d,q) and parameters every time a new
forecast is a better forecasting model (MAPE for the second model is
3.34% compared to 10.25% for the first model)

I got confused on understanding why the model performs better with the
method 2 or Am I doing something wrong? How do we explain this model in
this manufacturing case? The breakdown pattern of the machine changes
when repair work is done every time as breakdowns also depend on the
&quot;quality&quot; of the repair work. So it is hard to fix model order p,d,q
for the machine. Could this be a reason for the model 2 to perform
better? any thoughts? Also, is it practical and advisable to estimate
the model order everytime ?

**Comment by robjhyndman** [comment]
```header
Originally posted by Richard Lusch on 2017-06-05 13:58:46
````
Hello Professor,

How would you wrap a tslm model or linear model with arima errors into
this cross validation script? I\'ve been trying to work through it but
continue to run into errors.

Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-05 22:09:36
````
The tsCV function now does most of the work for your. See the help file.

**Comment by robjhyndman** [comment]
```header
Originally posted by Roy on 2017-06-20 06:37:07
````
Since the most recent updated version of the forecast package i\'ve been
getting an error message when attempting to include a regressor. This
error did not occur before.\
&quot;Error in if (ncol(xreg) != ncol(model\$xreg)) stop(&quot;Number of
regressors does not match fitted model&quot;) :\
argument is of length zero&quot;

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-21 00:11:25
````
Please provide a minimal reproducible example and post at
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/issues)

**Comment by robjhyndman** [comment]
```header
Originally posted by Naveena  on 2017-08-03 18:16:15
````
Hi Rob,

Thank you for your informative post. I wanted to apply to same to daily
data, but I am a little lost.\
My daily time series data starts at 2015-01-01 and ends at 2017-03-31.

Specific questions include:\
1.What should k be in my case?\
2.In mae1 \&lt;- mae2 \&lt;- mae3 \&lt;- matrix(NA,n-k-12,12), instead of 12 will
I use 365?\
3.In for(i in 1:(n-k-12)) will I use 365?

I have a very high level understanding of the code and I believe that is
another reason why I cannot translate this to a daily usecase. Any help
in clarifying the problem would be greatly appreciated!

**Comment by robjhyndman** [comment]
```header
Originally posted by William Ampeh on 2017-11-26 06:37:14
````
Dear Professor, I am currently looking into applying ML methods to time
series prediction, so your topic on CV and the later explanation of CV
as applied to time series prediction has been of great help. One
suggestion I have besides what has already been suggested by the others
is to enforce an upper bound for the &quot;window()&quot;. That is, to make sure
the &quot;window()&quot; does not look beyond the &quot;end date&quot; of series
&quot;a10&quot;. This will eliminate the &quot;warning messages&quot; (see attached
picture).
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/500208a16cd16073039b5b3b7fb80e0e113ef56844ca9e01c4841d15949e7850.png)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-11-26 12:04:41
````
Use the tsCV function instead.

**Reply by robjhyndman** [reply]
```header
Originally posted by William Ampeh on 2017-12-03 23:25:57
````
Thank you Professor Hyndman,

I have been struggling with the following 3 tasks:

\(1\) How to specify the window size\
tsCV(a10, fcast1, h=12, window=12) \# results in NA values\
\-\--

\(2\) How to incorporate k for k-Fold validation\
I am not sure how to work\
CVar(a10, k=5, lambda=0.15)

into tsCV(a10, fcast1, h=12, window=12)\
\-\--

\(3\) How do I produce values for a plot similar to what you have on
[https://robjhyndman.com/hyn\...](https://robjhyndman.com/hyndsight/tscvexample/)
?

I tried the following:\
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

fcast1 \&lt;- function(x, h){ forecast( tslm(x \~ trend + season,
lambda=0), h=h ) }\
fcast2 \&lt;- function(x, h){\
forecast( Arima(x, order=c(3,0,1), seasonal=list(order=c(0,1,1),
period=12),\
include.drift=TRUE, lambda=0, method=&quot;ML&quot;), h=h ) }\
fcast3 \&lt;- function(x, h){ forecast( ets(x, model=&quot;MMM&quot;,damped=TRUE),
h=h ) }

#\--\
mae1 \&lt;- tsCV(a10, fcast1, h=12) #mae1 \&lt;- tsCV(a10, fcast1, h=12,
window=12)\
mae2 \&lt;- tsCV(a10, fcast2, h=12)\
mae3 \&lt;- tsCV(a10, fcast3, h=12)

#\--\
plot(1:12, colMeans( matrix(mae1, ncol=12, byrow = F), na.rm=TRUE),
type=&quot;l&quot;, col=2,\
xlab=&quot;horizon&quot;, ylab=&quot;MAE&quot;, ylim=c(-3.0, 3.0))\
lines(1:12, colMeans( matrix(mae2, ncol=12, byrow = F), na.rm=TRUE),
type=&quot;l&quot;,col=3)\
lines(1:12, colMeans (matrix(mae3, ncol=12, byrow = F), na.rm=TRUE),
type=&quot;l&quot;,col=4)\
legend(&quot;topleft&quot;,legend=c(&quot;LM&quot;,&quot;ARIMA&quot;,&quot;ETS&quot;),col=2:4,lty=1)

#\-\--

\# This produced the following plot:

[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/ae0d803b1e7e7e31d5d737cf4e1f5be36501559a156f22e61331119dfcaab9b5.png)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-12-04 00:27:06
````
1\. Use a window big enough to fit the model.\
2. tsCV does not use k-fold CV. See this post:
[https://robjhyndman.com/hyn\...](https://robjhyndman.com/hyndsight/tscv/)\
3. tsCV does not use a minimum training window. To replicate the post,
throw away the first 60 values of the cv errors.

**Reply by robjhyndman** [reply]
```header
Originally posted by William Ampeh on 2017-12-04 07:26:45
````
Thank you Professor. I am still not getting the same results, please
where am I going wrong?

#\-- // Copy of code provided below, with plot attached.

#Define functions for tsCV\
fcast1 \&lt;- function(x, h){ forecast( tslm(x \~ trend + season,
lambda=0), h=h ) }\
fcast2 \&lt;- function(x, h){\
forecast( Arima(x, order=c(3,0,1), seasonal=list(order=c(0,1,1),
period=12),\
include.drift=TRUE, lambda=0, method=&quot;ML&quot;), h=h ) }\
fcast3 \&lt;- function(x, h){ forecast( ets(x, model=&quot;MMM&quot;,damped=TRUE),
h=h ) }\
#\--\
library(fpp) \# To load the data set a10

n \&lt;- length(a10);\
h =12\
#-\
mae1 \&lt;- tsCV(a10, fcast1, h=h)\
mae2 \&lt;- tsCV(a10, fcast2, h=h)\
mae3 \&lt;- tsCV(a10, fcast3, h=h);\
#\--\
keep_data \&lt;- n - 60

plot(1:12, colMeans( matrix( mae1 %\&gt;% tail(n=keep_data ), ncol=12,
byrow = F), na.rm=TRUE), type=&quot;l&quot;, col=2,\
xlab=&quot;horizon&quot;, ylab=&quot;MAE&quot;, ylim=c(0.65,1.05))\
lines(1:12, colMeans( matrix(mae2 %\&gt;% tail(n=keep_data ), ncol=12,
byrow = F), na.rm=TRUE), type=&quot;l&quot;,col=3)\
lines(1:12, colMeans (matrix(mae3%\&gt;% tail(n=keep_data ), ncol=12, byrow
= F), na.rm=TRUE), type=&quot;l&quot;,col=4)\
legend(&quot;topleft&quot;,legend=c(&quot;LM&quot;,&quot;ARIMA&quot;,&quot;ETS&quot;),col=2:4,lty=1,pch=21,
cex=0.4)

#\--// end of code
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/f8949e78d509975f504175d2ec6bb67d67336b69c6759ffcae094a6f9bdc6d0d.png)

**Comment by robjhyndman** [comment]
```header
Originally posted by Magnus AL on 2018-03-14 08:28:22
````
Hi. Im writing my master thesis on prediction of log salmon prices using
PLS-regression. The model uses both lagged values of salmon prices and
lagged values of other time series, i.e the model is multivariate. I am
considering using a normal k-fold CV for validation, but i am uncertain
if it is valid to use as the time series model is multivariate. I saw
you published a paper on the use of K-fold CV on purely autoregressive
time series models, but is it OK to use k-fold on multivariate lagged
time series models? I am assuming the residuals of the model are
stationary.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-03-14 20:22:50
````
Although our paper considers univariate AR models, it extends naturally
to multivariate AR models.

**Comment by robjhyndman** [comment]
```header
Originally posted by Cyrus Safaie on 2018-03-17 02:29:13
````
Thanks for the post. Have you thought through how this could be
efficiently scaled to 1000+ series?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-03-17 06:59:44
````
No

**Reply by robjhyndman** [reply]
```header
Originally posted by Cyrus Safaie on 2018-03-18 19:25:20
````
Thank you

**Comment by robjhyndman** [comment]
```header
Originally posted by Liliana on 2019-03-22 20:12:11
````
Hello Professor, Knowing that the sales volume increases per year, as
clearly shown here
[https://otexts.com/fpp2/sea\...](https://otexts.com/fpp2/seasonal-plots.html; &quot;https://otexts.com/fpp2/seasonal-plots.html;&quot;)
in the k-fold cross-validation example, the mae error increases as the
years increase (approx). Wouldn\'t it be better to scale the error based
on the year average, or use &quot;mean absolute scaled error&quot; MASE instead?
Thanks

**Comment by Charles-cod22** [comment]
Hello Professor,

I have some question on VAR cross validation.
Please, I have 2 variables in VAR model with data time length T = 30.
I find it a bit difficult to implement the cross validation method as you have it in your paper "Measuring forecast accuracy".

Could you clarify if it is possible to do CV for VAR model? Thank you.

Best,

Charles.

**Reply by robjhyndman** [reply]
Here's an example of time series cross validation using a VAR(3):

``` r
library(fpp3)
#> Registered S3 method overwritten by 'tsibble':
#>   method               from 
#>   as_tibble.grouped_df dplyr
#> ── Attaching packages ──────────────────────────────────────────── fpp3 1.0.2 ──
#> ✔ tibble      3.3.0     ✔ tsibble     1.1.6
#> ✔ dplyr       1.1.4     ✔ tsibbledata 0.4.1
#> ✔ tidyr       1.3.1     ✔ feasts      0.4.2
#> ✔ lubridate   1.9.4     ✔ fable       0.4.1
#> ✔ ggplot2     4.0.1
#> ── Conflicts ───────────────────────────────────────────────── fpp3_conflicts ──
#> ✖ lubridate::date()    masks base::date()
#> ✖ dplyr::filter()      masks stats::filter()
#> ✖ tsibble::intersect() masks base::intersect()
#> ✖ tsibble::interval()  masks lubridate::interval()
#> ✖ dplyr::lag()         masks stats::lag()
#> ✖ tsibble::setdiff()   masks base::setdiff()
#> ✖ tsibble::union()     masks base::union()

lung_deaths <- cbind(mdeaths, fdeaths) |>
  as_tsibble(pivot_longer = FALSE)
lung_stretch <- lung_deaths |>
  stretch_tsibble(.init = 36, .step = 1)

fit <- lung_stretch %>%
  model(VAR(vars(mdeaths, fdeaths) ~ AR(3)))
fit |>
  forecast(h=1) |>
  accuracy(lung_deaths)
#> Warning: The future dataset is incomplete, incomplete out-of-sample data will be treated as missing. 
#> 1 observation is missing at 1980 Jan
#> # A tibble: 2 × 11
#>   .model       .response .type     ME  RMSE   MAE   MPE  MAPE  MASE RMSSE   ACF1
#>   <chr>        <chr>     <chr>  <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>  <dbl>
#> 1 VAR(vars(md… fdeaths   Test  20.3    90.6  66.1  2.57  11.3 0.972 0.845 -0.144
#> 2 VAR(vars(md… mdeaths   Test   0.513 256.  189.  -1.66  12.8 1.08  0.971 -0.135
```

<sup>Created on 2025-12-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

It's much easier using the `fable` and `tsibble` packages (loaded with `fpp3`). See https://otexts.com/fpp3/tscv.html for further information.

**Comment by Charles-cod22** [comment]
Hello Professor,

Thank you very much for the example and clarification.

Best regards,

Charles.
